src/definitions.rs

```diff
@@ -92,114 +102,391 @@ impl From<FieldPath> for RichTerm {
+                    .map(|s| CurrentSchema::Schema(s)),
+                SchemaPointerElt::Else => current
+                    .object()?
+                    .subschemas
+                    .as_ref()?
+                    .else_schema
+                    .as_ref()
+                    .map(|s| CurrentSchema::Schema(s.as_ref())),
+            }?;
+
+            current = new;
+        }
 
-    /// Tries to interpret `self` as pointing to a top-level definition. A JSON pointer points to a
-    /// top-level definition if the path has exactly two elements and the first one is
-    /// `definitions`.
-    fn try_as_def(&self) -> Option<String> {
-        if self.path.len() == 2 && self.path[0] == "definitions" {
-            Some(self.path[1].clone())
+        if let CurrentSchema::Schema(current) = current {
+            Some(current)
         } else {
             None
         }
     }
 }
 
-/// The conversion of a JSON schema definition into a Nickel predicate and contract. We don't
-/// always use both, so we only store the part which is actually used.
-#[derive(Clone)]
-pub struct ConvertedDef {
-    doc: Option<Documentation>,
-    predicate: Option<Predicate>,
-    contract: Option<Contract>,
-}
+impl std::fmt::Display for SchemaPointer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for elt in self.path.iter() {
+            write!(f, "/{elt}")?;
+        }
 
-impl ConvertedDef {
-    /// Take the contract part out of this definition and convert it to a record field with the
-    /// appropriate definition. This method returns `None` if `self.contract` is `None`.
-    ///
-    /// After calling this method, `self.contract` will be `None`.
-    pub fn contract_as_field(&mut self) -> Option<Field> {
-        Self::as_field(self.contract.take(), self.doc.clone())
+        Ok(())
     }
+}
 
-    /// Take the predicate part out of this definition and convert it to a record field with the
-    /// appropriate definition. This method returns `None` if `self.contract` is `None`.
-    ///
-    /// After calling this method, `self.predicate` will be `None`.
-    pub fn predicate_as_field(&mut self) -> Option<Field> {
-        Self::as_field(self.predicate.take(), self.doc.clone())
+/// Element of a JSON pointer supported by json-schema-to-nickel. A JSON Pointer is parsed as a
+/// sequence of [SchemaPointerElt] which locates a subschema within a JSON Schema. Each variant
+/// corresponds to a JSON schema keyword.
+///
+/// Keywords containing a single schema (such as `contains`) don't need any additional data.
+/// Keywords storing objects carries a string indicating which property of the object should be
+/// accessed (e.g. `properties/foo`). Keywords storing arrays carries the index of the array to
+/// access (e.g. `items/0`).
+///
+/// `prefixItems` is a JSON Schema keyword that could be supported as well but it's unfortunately
+/// not supported by `schemars`, so we ignore it.
```

Comment:
We're accumulating a lot of quirks in `schemars`, it seems. Last I checked, `schemars` also isn't really maintained anymore, so maybe we should just start processing the JSON schemas ourselves.

Reply:
Yes, I agree. Although I don't know how much `prefixItems` is used, given that we just want to read structured schemas, this might not be too much of work. Another solution is to look for a better maintained fork or an alternative. Another more annoying issue this that schemars doesn't handle nested definitions, which do appear e.g. in the JSON schema test suite.